x-pack/plugins/security/server/lib/authorization/has_privileges.js

```diff
@@ -0,0 +1,107 @@
+  const versionPrivilege = getVersionPrivilege(kibanaVersion);
+
+  return function hasPrivilegesWithRequest(request) {
+
+    const hasApplicationPrivileges = async (privileges) => {
+      const privilegeCheck = await callWithRequest(request, 'shield.hasPrivileges', {
+        body: {
+          applications: [{
+            application,
+            resources: [DEFAULT_RESOURCE],
+            privileges
+          }]
+        }
+      });
+
+      const hasPrivileges = privilegeCheck.application[application][DEFAULT_RESOURCE];
+
+      // We include the login action in all privileges, so the existence of it and not the version privilege
+      // lets us know that we're running in an incorrect configuration. Without the login privilege check, we wouldn't
+      // know whether the user just wasn't authorized for this instance of Kibana in general
+      if (!hasPrivileges[getVersionPrivilege(kibanaVersion)] && hasPrivileges[getLoginPrivilege()]) {
+        throw new Error('Multiple versions of Kibana are running against the same Elasticsearch cluster, unable to authorize user.');
+      }
+
+      return {
+        username: privilegeCheck.username,
+        hasAllRequested: privilegeCheck.has_all_requested,
+        privileges: hasPrivileges
+      };
+    };
+
+    const hasPrivilegesOnKibanaIndex = async () => {
+      const privilegeCheck = await callWithRequest(request, 'shield.hasPrivileges', {
+        body: {
+          index: [{
+            names: [kibanaIndex],
+            privileges: ['create', 'delete', 'read', 'view_index_metadata']
+          }]
+        }
+      });
+
+      return Object.values(privilegeCheck.index[kibanaIndex]).includes(true);
+    };
+
+    return async function hasPrivileges(privileges) {
+      const allPrivileges = [versionPrivilege, loginPrivilege, ...privileges];
+      const privilegesCheck = await hasApplicationPrivileges(allPrivileges);
+
+      const username = privilegesCheck.username;
+
+      // We don't want to expose the version privilege to consumers, as it's an implementation detail only to detect version mismatch
+      const missing = Object.keys(privilegesCheck.privileges)
+        .filter(p => !privilegesCheck.privileges[p])
+        .filter(p => p !== versionPrivilege);
+
+      if (privilegesCheck.hasAllRequested) {
+        return {
+          result: HAS_PRIVILEGES_RESULT.AUTHORIZED,
+          username,
+          missing,
+        };
+      }
+
+      if (!privilegesCheck.privileges[loginPrivilege] && await hasPrivilegesOnKibanaIndex()) {
+        const msg = `Relying on implicit privileges determined from the index privileges is deprecated and will be removed in Kibana 7.0`;
+        server.log(['warning', 'deprecated', 'security'], msg);
+
+        return {
+          result: HAS_PRIVILEGES_RESULT.LEGACY,
+          username,
+          missing,
+        };
+      }
+
+      return {
+        result: HAS_PRIVILEGES_RESULT.UNAUTHORIZED,
```

Comment:
At risk of overthinking this initial design, wouldn't booleans make this easier?

```js
{
  authorized,
  legacy,
  username,
  missing
}
```

Then when consuming this, rather than string comparisons in a switch, you could do:

```js
const privilegeCheck = await checkPrivileges();
if (!privilegeCheck.authorized) {
  // audit log
  // reject early
}

// todo: remove in 7.0
if (privilegeCheck.legacy) {
  // audit log
  // callWithRequest
}

// internalRepository
```

This really isn't a big deal, so if you're not feeling it, don't bother. It would mean that errors and legacy case are handled early, and when the legacy case goes away, the code looks more like a standard fail-early then succeed flow, and you'd avoid the "this should never happen" default handler that comes with string comparison switch statements.